File: app/src/main/java/com/thoughtworks/china/mobile/collection/Collections.java
package com.thoughtworks.china.mobile.collection;


import com.google.common.base.Function;
import com.google.common.base.Predicate;
import com.google.common.collect.FluentIterable;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.collect.Lists.newArrayList;

public class Collections {

    public static void main(String[] args) {
        List<Integer> integers = newArrayList(1, 2, 3);
        integers.add(4);
        System.out.println(integers);

        ArrayList<Integer> range = newArrayList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
        FluentIterable<Integer> temps = FluentIterable.from(range)
                .filter(new Predicate<Integer>() {
                    @Override
                    public boolean apply(Integer input) {
                        return input % 2 == 1;
                    }
                })
                .transform(new Function<Integer, Integer>() {
                    @Override
                    public Integer apply(Integer input) {
                        return input * input;
                    }
                });
        int sum = 0;
        for (int temp : temps) {
            sum += temp;
        }
        System.out.println(sum);
    }
}

File: app/src/main/kotlin/com/thoughtworks/china/mobile/collection/Collections.kt
package com.thoughtworks.china.mobile.collection

object Main {

    @JvmStatic
    fun main(args: Array<String>) {
        val integers = mutableListOf(1, 2, 3)
        integers.add(4)

        (1..10)
                .filter { it % 2 == 1 }
                .map { it * it }
                .reduce { acc, i -> acc + i }
                .apply { println(this) }
    }
}
